### ISSUE TITLE
False negative `reportPossiblyUnboundVariable` for certain deletions in enclosing scope

### ISSUE THREAD

**Issue by dylwil3** [microsoft/pyright#9638]
A (non local) variable used in a function scope should be reported as possibly unbound if it is deleted anywhere in the enclosing scope. Currently this _is_ reported if the deletion occurs before the function definition, but that shouldn't affect the result (since function bodies are evaluated later).

Code sample in [pyright playground](https://pyright-play.net/?strict=true&enableExperimentalFeatures=true&reportImplicitOverride=true&reportShadowedImports=true&reportUnusedCallResult=true&code=CYUwZgBGAUCUBcAoCKIA8IF4IEZmtEgHM4lVyIAHAJwEsA7AF2jVggGIJGALWgZwh9uAewCuAG2AQARiAjUQlYdUYgpAQwFK%2BfWtPEBPCKPrSx9YPhShx6K-JCNR1ehCKIgA)

```python
def f():
    x = 1
    def g():
        print(x) # this should be reported as possibly unbound
    del x
    return g

```
Note that I am not asking for the much harder static analysis problem of context sensitivity for function calls - e.g. I think one should _also_ report `x` as possibly unbound in both of the following situation, even though the first is technically a false positive:

```python
def f_1():
    x = 1
    def g():
        print(x) # this should be reported as possibly unbound
    g()
    del x
    return 17

def f_2():
    x = 1
    def g():
        print(x) # this should be reported as possibly unbound
    g()
    del x
    return 17

```

More generally, it seems correct to report the following, but perhaps I'm missing some subtlety about module level scopes and imports?

```python
x = 1
del x
def g():
    print(x) # this should be reported as possibly unbound

```
(it's not _definitely_ unbound, since someone could import the module and redefine `x`.)

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug. I'll recast it as an enhancement request and consider it in that light.

I don't think this is a common source of bugs, so I'm reluctant to invest the time to implement this feature unless/until I receive signal from more pyright users that it would be useful. I'm going go reject the feature request for now, but I'm willing to revisit this decision in the future if this issue receives sufficient upvotes.
